### PR TITLE
Improve documentation for `ExtraCoords` methods and `physical_types`

### PR DESCRIPTION
**Pull request by dhomeier**
### Description
This is addressing some of the open documentation issues, specifically #444 and other `TableCoordinate`-related interfaces discussed there.

First, this adds `ndcube.extra_coords.table_coord` to the API reference, as a proxy for `BaseTableCoordinate` and to provide links to its subclasses. Also trying to fix some other broken links in `tabular_coordinates.rst`.

`physical_types` is explicitly declared and documented as kwarg in `ExtraCoords.add` and `ExtraCoords.from_lookup_tables`; as a by-product the latter now also passes `physical_types` on.

Some of the x-refs with labels look a bit unsightly in the inline docstrings, but should render decently in the built docs.

I was trying to also label the intersphinx link for gwcs somehow like `gwcs <~gwcs>`, but that is not recognised as external reference – if anyone has an idea how to do this, or what the intersphinx reference to the toplevel gwcs docs would be, pointers welcome. But since it's only in a single place, putting in the explicit URL is probably OK.

### CODE REVIEW

**Comment by dhomeier** [comment]
astropy needed a new `devdeps` version since 4.3x has been released. :)

File: ndcube/extra_coords/extra_coords.py [hunk 1]
```diff
@@ -38,23 +38,30 @@ class ExtraCoordsABC(abc.ABC):
...
-        lookup_table
-            The lookup table. Note, if this table is multi-dimensional it must
-            (currently) be specified with its axes in world order, so
-            transposed with respect to the data array.
+        lookup_table : `object` or sequence of `object`
+            The lookup table. A `BaseTableCoordinate <.table_coord>` subclass or anything
+            that can instantiate one, i.e. currently a `~astropy.time.Time`,
+            `~astropy.coordinates.SkyCoord`, or a (sequence of) `~astropy.units.Quantity`.
+            Note, if this table is multi-dimensional it must (currently) be specified with
+            its axes in world order, so transposed with respect to the data array.
```
Comment (DanRyanIrish) [inline ndcube/extra_coords/extra_coords.py:60]: I don't think this is true any more.  I think the axis is now given in array order.

**Comment by dhomeier** [comment]
Is there a `451.documentation.rst` category for the changelog? Though technically with the added support for `physical_types` in `from_lookup_tables` this might also go under `feature`.

**Review by DanRyanIrish — COMMENTED** [review]
Hi @dhomeler.  This looks like a great help.  I have one clarification about the ordering of `axis` in `ExtraCoords.add`.

File: ndcube/extra_coords/extra_coords.py (hunk 1 above)
Comment (dhomeier) [reply ndcube/extra_coords/extra_coords.py:60]: So just cut the last sentence, or better explicitly state that it has changed from 1.x?

**Comment by DanRyanIrish** [comment]
> Is there a `451.documentation.rst` category for the changelog? Though technically with the added support for `physical_types` in `from_lookup_tables` this might also go under `feature`.

You should be able to use `451.doc.rst`.  I think a `doc` type changelog is fine here since the main thrust of the PR is making how to use `physical_types` clearer.

File: ndcube/extra_coords/extra_coords.py (hunk 1 above)
Comment (dhomeier) [reply ndcube/extra_coords/extra_coords.py:60]: In #416 it sounds as if at this point there is no support for multi-dimensional tables at all.

Comment (DanRyanIrish) [reply ndcube/extra_coords/extra_coords.py:60]: Ah, I slightly misread this sentence initially which has caused a little confusion.  There is support for multi-coordinate tables.  But each individual coordinate must be 1D and can only correspond to one array axis.

On a separate point that caused my initial confusion, the ordering in which the single array axis must be specified is in array order, not pixel order.  For example, say you have a 3-D cube and want to add a 1-D extra coord to the 0th array axis.  In pixel order you would give this as axis 2 because pixel and array ordering are reversed.  However, the API for `ExtraCoords.add` requires you use array ordering to specify the single axis, i.e. 0.  I now realise you were not referring to this part of the API, but my misunderstanding prompted my response.

So to summarise, just cut the last sentence.

**Comment by DanRyanIrish** [comment]
> > Is there a `451.documentation.rst` category for the changelog? Though technically with the added support for `physical_types` in `from_lookup_tables` this might also go under `feature`.
> 
> You should be able to use `451.doc.rst`. I think a `doc` type changelog is fine here since the main thrust of the PR is making how to use `physical_types` clearer.

Alternatively, you can add two changelogs, `451.doc.rst` and `451.feature.rst` and break up the descriptions of the two different types of changes.  This might actually be the best way.

**Comment by Cadair** [comment]
> I was trying to also label the intersphinx link for gwcs somehow like gwcs <~gwcs>, but that is not recognised as external reference – if anyone has an idea how to do this, or what the intersphinx reference to the toplevel gwcs docs would be, pointers welcome.

I think I have run into this before, I don't think there is a high level reference. (Perhaps we should PR one?!)

**Comment by Cadair** [comment]
Looks good to me :+1:

File: ndcube/extra_coords/extra_coords.py (hunk 1 above)
Comment (dhomeier) [reply ndcube/extra_coords/extra_coords.py:60]: I have tried not to touch the actual content of the existing description of lookup tables, so did not check if any of that had changed recently. But if that bit is obsolete now I'll cut it.

**Comment by dhomeier** [comment]
I wasn't sure if the gwcs documentation is still under rapid development and correspondingly fluid...
We could probably address this on our side with ReST substitutions as astropy has them in `common_links.txt`; don't know if that's worth it just for gwcs and perhaps BaseTableCoordinate at this point.

**Comment by dhomeier** [comment]
I should add that in the test I added I haven't spent much thought on whether this is a sensible application of `from_lookup_tables`, as I initially expected `ExtraCoords.from_lookup_tables(["time", "exposure_time"], ...)` would be strictly equivalent to the successive `ec.add('time', ...); ec.add('exposure time', ...)` in the previous test before realising that the former is a class method.
So in the new test example checking mapping will raise
```python
>>> ec.mapping
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/derek/git/ndcube/ndcube/extra_coords/extra_coords.py", line 250, in mapping
    converter = partial(convert_between_array_and_pixel_axes, naxes=len(self._ndcube.dimensions))
AttributeError: 'NoneType' object has no attribute 'dimensions'
```
Frankly I didn't fully understand (quite literally) the Magic going on when initialising `ExtraCoords(cube)` in the first test, or how to correctly reproduce this using `from_lookup_tables`. The NDCube can be associated with it by `ec._ndcube = cube`, making this work in a somewhat after the fact manner; maybe the more straightforward use would be for an existing
```python
my_cube._extra_coords = ExtraCoords.from_lookup_tables(['time', 'exp_time'], (0, 0), [input_time, exp_time], physical_types=['custom:time:creation', 'custom:time:duration'])
```
but I haven't found a way to do this on initialisation or using public methods.

For checking the physical types the test works so far, but since there is only one other test for `from_lookup_tables`, and that's testing an exception, it could perhaps use some more coverage and documentation how to use it in real code.

**Comment by dhomeier** [comment]
I've put the `mapping` check back in a `pytest.raises` so it's not forgotten; otherwise I think the physical types are now addressed and the general `from_lookup_tables` usage can go into a separate issue, if necessary.

**Review by DanRyanIrish — APPROVED** [review]

**Comment by DanRyanIrish** [comment]
Hi @dhomeier.  This looks great.  I think it's ready to merge.  Does anyone else have any final requests?  @Cadair?
